src/clients/HubPoolClient.ts

```diff
@@ -345,6 +347,188 @@ export class HubPoolClient extends BaseAbstractClient {
+   * @param timeToCache Age at which the response is able to be cached.
+   * @returns HubPool utilization at `blockNumber` after optional `amount` increase in utilization.
+   */
+  protected async _getUtilization(
+    hubPoolToken: string,
+    blockNumber: number,
+    amount: BigNumber,
+    timestamp: number,
+    timeToCache: number
+  ): Promise<BigNumber> {
+    // Resolve this function call as an async anonymous function
+    const resolver = async () => {
+      const overrides = { blockTag: blockNumber };
+      if (amount.eq(0)) {
+        // For zero amount, just get the utilisation at `blockNumber`.
+        return await this.hubPool.callStatic.liquidityUtilizationCurrent(hubPoolToken, overrides);
+      }
+
+      return await this.hubPool.callStatic.liquidityUtilizationPostRelay(hubPoolToken, amount, overrides);
+    };
+
+    // Resolve the cache locally so that we can appease typescript
+    const cache = this.cachingMechanism;
+
+    // If there is no cache or the timestamp is not old enough to be cached, just resolve the function.
+    if (!cache || !shouldCache(getCurrentTime(), timestamp, timeToCache)) {
+      return resolver();
+    }
+
+    // Otherwise, let's resolve the key
+    const key = amount.eq(0)
+      ? `utilization_${hubPoolToken}_${blockNumber}`
+      : `utilization_${hubPoolToken}_${blockNumber}_${amount.toString()}`;
+    const result = await cache.get<string>(key);
+    if (isDefined(result)) {
+      return BigNumber.from(result);
+    }
+
+    // We were not able to find a valid result, so let's resolve the function.
+    const utilization = await resolver();
+    if (cache && shouldCache(getCurrentTime(), timestamp, timeToCache)) {
+      // If we should cache the result, store it for up to DEFAULT_CACHING_TTL.
+      await cache.set(key, `${utilization.toString()}`, DEFAULT_CACHING_TTL);
+    }
+
+    return utilization;
+  }
+
+  async batchComputeRealizedLpFeePct(
+    deposits: Pick<
+      DepositWithBlock,
+      "quoteTimestamp" | "amount" | "originChainId" | "originToken" | "destinationChainId" | "blockNumber"
+    >[]
+  ): Promise<{ quoteBlock: number; realizedLpFeePct?: BigNumber }[]> {
+    if (!isDefined(this.currentTime)) {
+      throw new Error("HubPoolClient has not set a currentTime");
+    }
+    const timeToCache = this.configOverride.timeToCache ?? DEFAULT_CACHING_SAFE_LAG;
+
+    // Map SpokePool tokens to HubPool tokens, and each HubPool token to a quoteTimestamp.
+    // This will later be reconciled to a HubPool blockNumber.
+    const hubPoolTokens: { [originToken: string]: string } = {};
+    const utilizationTimestamps: { [hubPoolToken: string]: number[] } = {};
+    const { originChainId: firstOriginChainId } = deposits[0];
+    deposits.forEach((deposit) => {
+      const { originChainId, originToken, quoteTimestamp } = deposit;
+      assert(
+        originChainId === firstOriginChainId,
+        "Cannot compute bulk realizedLpFeePct for disparate origin chains" +
```

Comment:
```suggestion
        "Cannot compute bulk realizedLpFeePct for different origin chains" +
```

Is "different" more precise of a word than "disparate"?